pkg/processor/runtime/rpc/runtime.go

```diff
@@ -90,67 +94,89 @@ func NewRPCRuntime(logger logger.Logger, configuration *runtime.Configuration, r
-	} else {
-		listener, address, err = newRuntime.createTCPListener()
-	}
-
-	if err != nil {
-		return nil, errors.Wrap(err, "Can't create listener")
-	}
-
-	wrapperProcess, err := runWrapper(address)
-	if err != nil {
-		return nil, errors.Wrap(err, "Can't run wrapper")
-	}
-	newRuntime.wrapperProcess = wrapperProcess
-
-	conn, err := listener.Accept()
-	if err != nil {
-		return nil, errors.Wrap(err, "Can't get connection from wrapper")
+	if err := newRuntime.startWrapper(); err != nil {
+		return nil, errors.Wrap(err, "Can't start wrapper")
 	}
-	newRuntime.Logger.Info("Wrapper connected")
-
-	newRuntime.eventEncoder = NewEventJSONEncoder(newRuntime.Logger, conn)
-	newRuntime.outReader = bufio.NewReader(conn)
 
 	newRuntime.SetStatus(status.Ready)
-
 	return newRuntime, nil
 }
 
 // ProcessEvent processes an event
 func (r *Runtime) ProcessEvent(event nuclio.Event, functionLogger logger.Logger) (interface{}, error) {
-	// TODO: Check that status is Ready?
+	if currentStatus := r.GetStatus(); currentStatus != status.Ready {
+		return nil, errors.Errorf("Processor not ready (current status: %s)", currentStatus)
+	}
+
 	r.Logger.DebugWith("Processing event",
 		"name", r.configuration.Meta.Name,
 		"version", r.configuration.Spec.Version,
 		"eventID", event.GetID())
 
-	resultChan := make(chan *result)
-	go r.handleEvent(functionLogger, event, resultChan)
-
-	select {
-	case result := <-resultChan:
-		r.Logger.DebugWith("Event executed",
-			"name", r.configuration.Meta.Name,
-			"status", result.StatusCode,
-			"eventID", event.GetID())
-		return nuclio.Response{
-			Body:        result.DecodedBody,
-			ContentType: result.ContentType,
-			Headers:     result.Headers,
-			StatusCode:  result.StatusCode,
-		}, nil
-	case <-time.After(eventTimeout):
-		return nil, fmt.Errorf("handler timeout after %s", eventTimeout)
+	r.functionLogger = functionLogger
+	// We don't use defer to reset r.functionLogger since it decreases performance
+	if err := r.eventEncoder.Encode(event); err != nil {
+		r.functionLogger = nil
+		return nil, errors.Wrapf(err, "Can't encode event: %+v", event)
+	}
+
+	result, ok := <-r.resultChan
+	r.functionLogger = nil
+	if !ok {
+		msg := "Client disconnected"
+		r.Logger.Error(msg)
+		r.SetStatus(status.Error)
+		r.functionLogger = nil
+		return nil, errors.New(msg)
+	}
+
+	return nuclio.Response{
+		Body:        result.DecodedBody,
+		ContentType: result.ContentType,
+		Headers:     result.Headers,
+		StatusCode:  result.StatusCode,
+	}, nil
+}
+
+// Stop stops the runtime
+func (r *Runtime) Stop() error {
+	if r.wrapperProcess != nil {
+		err := r.wrapperProcess.Kill()
+		if err != nil {
+			r.SetStatus(status.Error)
+			return errors.Wrap(err, "Can't kill wrapper process")
+		}
+	}
+
+	r.SetStatus(status.Stopped)
+	return nil
+}
+
+// Restart restarts the runtime
+func (r *Runtime) Restart() error {
+	if err := r.Stop(); err != nil {
+		return err
+	}
+
+	// Send error for current event
```

Comment:
We shouldn't blindly shove a timeout result when asked to restart. 
1. There might be a restart request unrelated to a time out
2. There might not even be a request pending a result

Reply:
The reasoning is:
1. We have no current request - the message will just be ignored
2. We have another request which is not filled - we'll get an error (wrong kind) since we're restarting

Another solution will be to add event ID to every message from/to the wrapper. Which means changing all the RPC runtimes.

Reply:
OK, seems reasonable